app/models/user.rb

```diff
@@ -70,55 +41,92 @@ def switch_to_idp!
 
   ## Trackable
   field :idp_uuid, type: String
-
-  field :roles, :type => Array, :default => []
-
-  # Oracle Identity Manager ID
-  field :oim_id, type: String, default: ""
-
-  field :last_portal_visited, type: String
   field :idp_verified, type: Boolean, default: false
 
-  index({preferred_language: 1})
-  index({approved: 1})
-  index({roles: 1},  {sparse: true}) # MongoDB multikey index
-  index({email: 1},  {sparse: true, unique: true})
-  index({oim_id: 1}, {sparse: true, unique: true})
-  index({created_at: 1 })
-
-  track_history :on => [:oim_id, :email],
-                :modifier_field => :modifier,
-                :modifier_field_optional => true,
-                :version_field => :tracking_version,
-                :track_create  => true,
-                :track_update  => true,
-                :track_destroy => true
-
-  before_save :strip_empty_fields
+  index({ account_id: 1 }, { sparse: true, unique: true })
+  index({ oim_id: 1 }, { sparse: true, unique: true })
+  index({ email: 1 }, { sparse: true, unique: true })
+  index({ preferred_language: 1 })
+  index({ approved: 1 })
+  index({ roles: 1 }, { sparse: true }) # MongoDB multikey index
+  index({ created_at: 1 })
+
+  track_history on: %i[oim_id email],
+                modifier_field: :modifier,
+                modifier_field_optional: true,
+                version_field: :tracking_version,
+                track_create: true,
+                track_update: true,
+                track_destroy: true
+
+  track_history on: %i[account_id email],
+                modifier_field: :modifier,
+                modifier_field_optional: true,
+                version_field: :tracking_version,
+                track_create: true,
+                track_update: true,
+                track_destroy: true
 
   # Enable polymorphic associations
   belongs_to :profile, polymorphic: true, optional: true
 
   has_one :person, inverse_of: :user
-  accepts_nested_attributes_for :person, :allow_destroy => true
+  accepts_nested_attributes_for :person, allow_destroy: true
 
-  attr_accessor :invitation_id
   #  validate :ensure_valid_invitation, :on => :create
 
+  validates_presence_of :oim_id
+  validates_uniqueness_of :oim_id, case_sensitive: false
+
+  validate :oim_id_rules
+
+  validates :email,
+            uniqueness: {
+              case_sensitive: false
+            },
+            format: {
+              with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i,
```

Comment:
this regex is not valid for emails since `mark@localhost` is valid, but this regex would say that was invalid